Update/edit flow for blocks which have a media placeholder

draganescu opened the pull request:
## Description
Closes #14795 

This is a follow up to the image block edit flow update which ports the new flow to all the blocks which have media with an edit state. The blocks affected are:

- media+text
- cover
- audio
- video
- file

## How has this been tested?
For now I've only tested locally.

## Types of change
New feature (non-breaking change which adds functionality)

## Screenshots

### Media + text
![media+text](https://user-images.githubusercontent.com/107534/55975613-0bd58000-5c93-11e9-9cd2-0dfca34fc26f.gif)

### Cover
![cover](https://user-images.githubusercontent.com/107534/55975900-a33ad300-5c93-11e9-92f5-71f86914cf64.gif)

### Audio
![audio](https://user-images.githubusercontent.com/107534/55976121-16444980-5c94-11e9-8858-a3d80ac636ef.gif)

### File
![file](https://user-images.githubusercontent.com/107534/55976214-53a8d700-5c94-11e9-98ee-a41235cc8466.gif)

### Video
![video](https://user-images.githubusercontent.com/107534/55976311-8783fc80-5c94-11e9-9262-1ac1cb55d890.gif)

### Embed
![embed](https://user-images.githubusercontent.com/107534/55976429-daf64a80-5c94-11e9-8ebc-e95a552e8f8f.gif)

draganescu committed:
adds the new edit media icon to the media-text, cover, audio, file and video blocks

draganescu committed:
add new placeholder to edit state of media text

draganescu committed:
fixes preview for media other than image

draganescu committed:
adds the new flow to the cover block

draganescu committed:
adds correct state to edit button, adds new flow to audio

draganescu committed:
fixed state for newly inserted audio, adds new flow to file

draganescu committed:
adds aria pressed to edit button, adds the new flow to video

gziolo commented on packages/block-library/src/video/edit.js:
```diff
@@ -150,22 +158,37 @@ class VideoEdit extends Component {
 			this.setState( { src: media.url, editing: false } );
 		};
 
+		const editImageIcon = ( <SVG width={ 20 } height={ 20 } viewBox="0 0 20 20"><Rect x={ 11 } y={ 3 } width={ 7 } height={ 5 } rx={ 1 } /><Rect x={ 2 } y={ 12 } width={ 7 } height={ 5 } rx={ 1 } /><Path d="M13,12h1a3,3,0,0,1-3,3v2a5,5,0,0,0,5-5h1L15,9Z" /><Path d="M4,8H3l2,3L7,8H6A3,3,0,0,1,9,5V3A5,5,0,0,0,4,8Z" /></SVG> );
```
Isn’t it the same icon duplicated in all components?

draganescu replied:
:( yes it is :D that's the kind of stuff I put WIP in the title for. On the other hand, in this particular case, these all are in the block-library package, where do you think is the most appropriate place to put this icon so it's accessible to all blocks?

gziolo replied:
@jasmussen - is it something that should be moved to `Dashicons` component? I think this is what we did in the past several times.

kjellr replied:
I believe the Dashicons component Gutenberg uses [comes directly from](https://github.com/WordPress/dashicons/blob/master/react/index.jsx) the Dashicons repo, which would mean we'd need to merge this icon there first. There've been lots of discussions around abandoning Dashicons in core, so I'm not really sure if that's what we want here. 🤔 Interested to hear what Joen thinks. 

Totally agree that we should put this in a central place though. Doesn't make sense to duplicate it so frequently.

jasmussen replied:
Apologize for missing this ping. I feel like Github notifications could do with some solid improvements, so actual mentions are unmissable.

The Dashicons component is created and maintained in a separate repo, and icons added there get a sanity check and are then merged and we copy the component here. It's not a great flow, and as Kjell mentioned we are looking at ways to smooth out this flow. 

In the mean time, I don't think we need to add this icon to Dashicons just to use it here. I think it's high time we look at a better flow, and I'm not sure it helps Dashicons to be overloaded with additional icons in the mean time. This is a personal opinion, but it's why I would suggest this icon shouldn't be a blocker.

draganescu committed:
adds the new edit flow to embed

kjellr commented:
This is looking great so far, @draganescu! I took it for a quick spin, and noticed just a couple small bugs: 

1. Would it be possible to add a "Cancel" button to the embeds? 

<img width="761" alt="Screen Shot 2019-04-11 at 12 02 35 PM" src="https://user-images.githubusercontent.com/1202812/55972945-88943a00-5c52-11e9-8900-9ff334b53635.png">

2. It looks like the image preview is a little misplaced in the Media + Text block. I think we need to adjust the margins + max-width there to be a bit more comfortable: 

<img width="724" alt="Screen Shot 2019-04-11 at 12 06 25 PM" src="https://user-images.githubusercontent.com/1202812/55973056-c85b2180-5c52-11e9-9eb2-c8079de2f522.png">

draganescu commented:
@kjellr I have implemented the flow to all the blocks which have this edit state, including those which were opening a media selection modal directly. I think this is better and makes block behaviour predictable.

So far there are two problems I know of:

- the file block can't work with an URL
- the embed block doesn't have a cancel link

Also I have marked this WIP b/c some of the code could be made more DRY and the edit/view pattern could be made the same in all blocks, maybe have an edit/view wrapper component?

draganescu committed:
removed duplicate definition

draganescu committed:
adds cancel link to embed edit mode

draganescu committed:
fixes the media preview in media + text block

draganescu committed:
new snapshot for the embed block

draganescu commented:
Rebased and updated with new snapshot so tests will pass. I think it's ready for review. Aside from code DRY-ness, there might be a point where we could make the code in these blocks more "standard" but maybe we could move ahead like this for now and refactor in a separate PR.

@kjellr if you can take it for another spin that'd be great! :)

kjellr commented:
Thanks, @draganescu! This is looking great. I noticed just one small bug: 

For embed blocks, the "Cancel" link appears even before you enter a URL: 

<img width="646" alt="Screen Shot 2019-04-18 at 12 20 32 PM" src="https://user-images.githubusercontent.com/1202812/56375774-6b241a80-61d4-11e9-9dc6-4bc17dcda417.png">

Once that's all set, I think this is all set from a design perspective. Before we decide to move ahead for sure though, I'd love to get @jasmussen's thoughts on [the DRY issue](https://github.com/WordPress/gutenberg/pull/14918#discussion_r274497548). He may have an idea for a quick fix.

jasmussen commented:
This is nice work, and it solves a problem, which is great. Like Kjell, I think it's close to shipping as an enhancement to what we have. There's just one issue that needs to be fixed before that:

![Screenshot 2019-04-19 at 08 23 44](https://user-images.githubusercontent.com/1204802/56410237-b134b900-627c-11e9-958c-f62d13454999.png)

In this screenshot the edit image button appears to be part of the alignment group. The edit image button should always be its own group, and never be part of any other groups, because its behavior is different.

---

All this said, I feel like this interface also starts to show some of the limitations of the current placeholder interface we have, and suggests that we should separately look at improving this in a holistic way.

- We have a generic appender for adding items, which features a dashed outline but same gray.
- There's no dashed outline on the media placeholder, but there's a whole slew of buttons that are increasingly unwieldy and missing and some layout. 
- The flow for going from final state and back to setup state is a little inconsistent.
- The button for going back to this state implies "Switch out", which is true, but it's still a little obtuse. And moreso, when it sits in the block toolbar it feels like it implies action on the whole block, which in the case of the Media & Text block is not the case. I wonder if it might be more clear to add the "Edit image" button to the overflow menu instead?
- Sometimes clicking the edit button shows a preview of what's currently applied (image), other times it doesn't (embeds). 

I don't feel like those issues are blockers for shipping this improvement. But it does feel like there is an opportunity to step back for a second and think: if we tackled appenders, edit buttons, and placeholder/setup states _together_, what could that do to improve the user experience? What could we eliminate? What could we improve?

draganescu committed:
removes the cancel link in embed initial state

draganescu committed:
moves the edit button to its own group, fixes media preview margin

draganescu commented:
rebased, fixed the button'd group (thanks @jasmussen), removed the cancel link (thanks @kjellr).
waiting on someone to review the code :)

draganescu committed:
new snapshot for the embed block

talldan commented:
This one might be easier to review (and ship) if it were one PR per block.

I'll post some observations from testing as a comment and then loop back to a code review:

**Media & Text**

- Some space is needed between the Media Library and Insert From URL buttons (seems to also apply to other blocks):
![Screen Shot 2019-05-02 at 5 13 06 pm](https://user-images.githubusercontent.com/677833/57066113-ea017300-6cfd-11e9-9b7b-59176ea4ed5e.png)

- I have access to the Insert from URL option when logged in as a contributor, but the image is not added when I use it (it appears blank):
![Screen Shot 2019-05-02 at 5 23 29 pm](https://user-images.githubusercontent.com/677833/57066559-149ffb80-6cff-11e9-9869-17660178ace5.png)

**Cover**

- As a contributor, the edit image toolbar button doesn't show up after inserting from a URL (seems to work nicely as an admin, though):
![Screen Shot 2019-05-02 at 5 26 40 pm](https://user-images.githubusercontent.com/677833/57066698-79f3ec80-6cff-11e9-9bc4-d5728b50539b.png)

**Audio**
- The media library seems to be able to get a 'poster' for an audio clip. A future improvement might be to see if we can also access this and use it as the preview on the placeholder. The same may also be true for videos:
![Screen Shot 2019-05-02 at 5 30 20 pm](https://user-images.githubusercontent.com/677833/57066902-1918e400-6d00-11e9-9543-34cbaaf631a1.png)

**File**
- I'm seeing an issue using the Upload button from the placeholder. The upload seems to take place but the block immediately shows the placeholder again.

- The filename and file link isn't updated when adding or replacing a file using 'Insert from URL' (could just use the url).

- As a contributor, there's no cancel button when replacing a file. (seems to be an issue for multiple blocks)

talldan commented on packages/block-library/src/audio/edit.js:
```diff
@@ -118,22 +122,35 @@ class AudioEdit extends Component {
+					<BlockControls>
+						{ !! src && ( <Toolbar>
+							<IconButton
+								className={ classnames( 'components-icon-button components-toolbar__control', { 'is-active': this.state.editing } ) }
```
No need to add the components-icon-button classname, this is added by the IconButton component itself.

talldan replied:
Also, it seems like `is-active` would always be true as this is within an if statement `if ( editing ) {`

talldan commented on packages/block-library/src/audio/edit.js:
```diff
@@ -143,7 +160,7 @@ class AudioEdit extends Component {
 							className="components-icon-button components-toolbar__control"
```
Same issue with the `components-icon-button` classname. Looks like it was an existing issue.

talldan replied:
Seems a shame to have to duplicate the code for this IconButton. Maybe it can be extracted to a separate component in the same file? Or the rendering logic could be reworked.

talldan commented on packages/block-library/src/audio/edit.js:
```diff
@@ -103,7 +107,7 @@ class AudioEdit extends Component {
 		const { setAttributes, isSelected, className, noticeOperations, noticeUI } = this.props;
 		const { editing } = this.state;
 		const switchToEditing = () => {
```
I also noticed this function doesn't `speak` when switching while some of the other blocks do.

talldan commented on packages/block-library/src/cover/edit.js:
```diff
@@ -225,25 +254,36 @@ class CoverEdit extends Component {
+						onDoubleClick={ this.toggleIsEditing }
+						onCancel={ !! url && this.toggleIsEditing }
+						notices={ noticeUI }
+						onError={ this.onUploadError }
```
Is this right? I couldn't see an onUploadError method.

talldan commented on packages/block-library/src/file/edit.js:
```diff
@@ -134,33 +166,58 @@ class FileEdit extends Component {
+					<BlockControls>
+						<Toolbar>
+							{ !! href && ( <IconButton
+								className={ classnames( 'components-icon-button components-toolbar__control', { 'is-active': this.state.isEditing } ) }
```
`components-icon-button` again

talldan commented on packages/block-library/src/file/edit.js:
```diff
@@ -134,33 +166,58 @@ class FileEdit extends Component {
+		const editImageIcon = ( <SVG width={ 20 } height={ 20 } viewBox="0 0 20 20"><Rect x={ 11 } y={ 3 } width={ 7 } height={ 5 } rx={ 1 } /><Rect x={ 2 } y={ 12 } width={ 7 } height={ 5 } rx={ 1 } /><Path d="M13,12h1a3,3,0,0,1-3,3v2a5,5,0,0,0,5-5h1L15,9Z" /><Path d="M4,8H3l2,3L7,8H6A3,3,0,0,1,9,5V3A5,5,0,0,0,4,8Z" /></SVG> );
 
-		if ( ! href || hasError ) {
+		if ( ! href || isEditing || hasError ) {
```
Might be able to drop `! href` here, since `isEditing` is now initialised to the right value.

talldan reviewed (commented):
I've made my way through about half the code, but there's quite a lot. I defo feel this could be split up into a PR per block. It'd be much easier to refine and ship each one indivdually.

One thing that having all the changes in one PR makes clear is how different yet the same all the media based blocks are. It'd be great to see some steps towards making the code the same across each one, even if it's just using the same style of code (methods vs inline functions, naming conventions).

Then it'd be easier at a later point to try to refactor them all to use some common code.

draganescu commented:
Thanks @talldan! It feels like a great moment to refactor some code indeed, my fear is that other PRs on these blocks plus refactoring will make some mega conflicts given time :D. I will try to split in more than one PR and see how that goes.

Not sure about the issues when logged in as a contributor and how the media placeholder could have caused them :/

talldan commented:
@draganescu - we can try to push this through for all the blocks, but based on experience I've seen that smaller PRs tend to ship more quickly and are more likely to be reviewed. They're also much easier to test.

Focussing on one block at a time is what I'd personally try to do.

A good way to do it might be to checkout the files from this branch for one block onto a separate branch. Haven't tried it but something like this might work to bring the changes across onto a new branch:
```
> git checkout master
> git pull
> git checkout -b update/edit-flow-for-audio-block
> git checkout update/edit-flow-for-media -- packages/block-library/src/audio/
```

draganescu commented:
@talldan as suggested I've split this into multiple PRs for easier review. I have also incorporated the bits collected from your past reviews (hopefully) in the new PRs.

Here they are:

## Audio Block
https://github.com/WordPress/gutenberg/pull/15520

## Cover Block
https://github.com/WordPress/gutenberg/pull/15519

## Embed block
https://github.com/WordPress/gutenberg/pull/15518

## Media + Text Block
https://github.com/WordPress/gutenberg/pull/15517

## Video block
https://github.com/WordPress/gutenberg/pull/15516

## File block
https://github.com/WordPress/gutenberg/pull/15515
